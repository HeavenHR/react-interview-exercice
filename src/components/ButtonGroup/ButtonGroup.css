.btn {
  text-transform: capitalize;
}

.btn-primary:active {
  background-color: #337ab7 !important;
  border-color: #2e6da4 !important;
}

.fa-male, .fa-female {
  padding-right: 10px;
}

.form-group:last-child {
  margin-bottom: 0;
}