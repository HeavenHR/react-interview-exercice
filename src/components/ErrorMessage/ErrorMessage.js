import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

const ErrorMessage = ({ message }) =>
  <span className={classnames('help-block')}>{message}</span>;

ErrorMessage.propTypes = {
  message: PropTypes.string.isRequired
};

export default ErrorMessage;
