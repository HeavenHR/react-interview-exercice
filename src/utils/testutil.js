import { mount, shallow } from 'enzyme';
import { prettyPrint } from 'html';
import * as types from '../constants/ActionTypes';

export const mockFunction = jest.fn(value => value);

export const sampleFriend = {
  name: 'Sample friend',
  gender: 'male'
};

export const sampleId = 1;

export const createMountedComponent = ({ children, isShallow }) => {
  const mountedComponent = isShallow ? shallow(children) : mount(children);
  mountedComponent.htmlRender = () => prettyPrint(mountedComponent.html(), { indent_size: 2 });

  return mountedComponent;
};

export const inputProps = {
  value: 'sample name',
  onSubmit: mockFunction,
  onChange: mockFunction,
  isInvalid: false,
  errorMessage: 'Please fill friend name'
};

export const buttonGroupProps = {
  selected: '',
  options: [
    { value: 'male', showIcon: true },
    { value: 'female', showIcon: true }
  ],
  onClick: mockFunction,
  isInvalid: false,
  errorMessage: 'Please select gender'
};

export const changeEventObject = { target: { value: 'sample text' } };

export const samplePages = [1, 2, 3, 4, 5, 6, 7];

export const totalPages = samplePages.length;

export const addFriendAction = {
  type: types.ADD_FRIEND,
  friend: sampleFriend
};

export const expectedAddFriendState = {
  friendsById: [
    {
      name: 'Sample friend',
      gender: 'male'
    },
    {
      name: 'Theodore Roosevelt',
      starred: true,
      gender: 'male'
    },
    {
      name: 'Abraham Lincoln',
      starred: false,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const deleteFriendAction = {
  type: types.DELETE_FRIEND,
  id: 0
};

export const expectedDeleteFriendState = {
  friendsById: [
    {
      name: 'Abraham Lincoln',
      starred: false,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const starFriendAction = {
  type: types.STAR_FRIEND,
  id: 1
};

export const expectedStarFriendState = {
  friendsById: [
    {
      name: 'Theodore Roosevelt',
      starred: true,
      gender: 'male'
    },
    {
      name: 'Abraham Lincoln',
      starred: true,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const updatePaginationAction = {
  type: types.UPDATE_PAGINATION,
  currentPage: 2
};

export const expectedUpdatedPaginationState = {
  pageSize: 2,
  currentPage: 2,
  pagingRange: 3
};
