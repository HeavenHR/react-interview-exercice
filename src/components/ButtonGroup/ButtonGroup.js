import React, { Component } from 'react';
import PropTypes from 'prop-types';
import className from 'classnames/bind';

import styles from './ButtonGroup.css';
import ErrorMessage from '../ErrorMessage';

const cn = className.bind(styles);

class ButtonGroup extends Component {
  static propTypes = {
    selected: PropTypes.string,
    options: PropTypes.arrayOf(PropTypes.object),
    onClick: PropTypes.func,
    isInvalid: PropTypes.bool,
    errorMessage: PropTypes.string
  }

  state = {
    selected: this.props.selected
  }

  componentWillReceiveProps(nextProps) {
    const { selected } = nextProps;

    if (this.props !== nextProps) {
      this.setState({ selected });
    }
  }

  handleClick = (selected) => () => {
    const { onClick } = this.props;

    this.setState({ selected })
    onClick && onClick(selected);
  }

  isActive = value => {
    return this.state.selected === value;
  }

  renderOptions = () => {
    const { options = [] } = this.props;
    
    // Note: Button should be a separate component, due to time constrain I used it directly here
    return options.map(({value, showIcon}) => (
      <button
        key={value}
        type="button"
        value={value}
        onClick={this.handleClick(value)}
        className={cn('btn btn-primary col-sm-6', { active: this.isActive(value) })}
      >
        {showIcon && <i className={cn(`fa fa-${value}`)} />}
        {value}
      </button>
    ));
  }

  render() {
    const { isInvalid, errorMessage } = this.props;
    return (
      <div className={cn('form-group', { 'has-error': isInvalid })}>
        <div className={cn('btn-group btn-block')} >
          {this.renderOptions()}
        </div>
        {isInvalid && errorMessage && <ErrorMessage message={errorMessage} />}
      </div>
    );
  }
}

export default ButtonGroup;
