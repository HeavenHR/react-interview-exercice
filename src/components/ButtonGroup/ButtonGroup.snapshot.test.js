import React from 'react';

import ButtonGroup from './ButtonGroup';
import { createMountedComponent, buttonGroupProps } from '../../utils/testutil';
import ErrorMessage from '../ErrorMessage';

describe('<ButtonGroup /> should', () => {
  let mountedButtonGroup;
  let buttons;

  beforeEach(() => {
    mountedButtonGroup = createMountedComponent({
      children: <ButtonGroup {...buttonGroupProps} />
    });

    buttons = mountedButtonGroup.find('button').map(item => item);
  });

  it('render correctly according to props', () => {
    const icons = mountedButtonGroup.find('i').map(item => item);

    expect(mountedButtonGroup.htmlRender()).toMatchSnapshot();
    expect(buttons.length).toEqual(2);
    expect(icons.length).toEqual(2);
    expect(icons[0].hasClass('fa fa-male')).toBeTruthy();
    expect(icons[1].hasClass('fa fa-female')).toBeTruthy();
    expect(mountedButtonGroup.state().selected).toEqual('');
  });

  it('render error message when has invalid prop', () => {
    const mountedButtonGroup = createMountedComponent({
      children: <ButtonGroup {...buttonGroupProps} isInvalid={true} />
    });

    expect(mountedButtonGroup.find('div.form-group').hasClass('has-error')).toBeTruthy();
    expect(mountedButtonGroup.find(ErrorMessage)).toHaveLength(1);
  });

  it('updates state and call onClick function on first button click', () => {
    buttons[0].simulate('click');

    expect(mountedButtonGroup.state().selected).toEqual('male');
    expect(mountedButtonGroup.props().onClick).toBeCalledWith('male');
  });

  it('updates state and call onClick function on second button click', () => {
    buttons[1].simulate('click');

    expect(mountedButtonGroup.state().selected).toEqual('female');
    expect(mountedButtonGroup.props().onClick).toBeCalledWith('female');
  });
});