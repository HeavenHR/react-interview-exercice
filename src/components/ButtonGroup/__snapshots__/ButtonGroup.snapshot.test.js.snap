exports[`<ButtonGroup /> should render correctly according to props 1`] = `
"<div class=\"form-group\">
  <div class=\"btn-group btn-block\">
    <button type=\"button\" value=\"male\" class=\"btn btn-primary col-sm-6\"><i class=\"fa fa-male\"></i>
      <!-- react-text: 5 -->male
      <!-- /react-text -->
    </button>
    <button type=\"button\" value=\"female\" class=\"btn btn-primary col-sm-6\"><i class=\"fa fa-female\"></i>
      <!-- react-text: 8 -->female
      <!-- /react-text -->
    </button>
  </div>
</div>"
`;
